### ISSUE TITLE
How to add `--cap-add` arguments

### ISSUE THREAD

**Issue by davidmpa** [caprover/caprover#683]
**IMPORTANT:**
Please quickly search the existing issues and make sure that you're not create a duplicate issue. If you have a quesetion about a particular Dockerfile and deployment, change your questions to Dockerfile and post it where appropriate such as StackOverflow. This is a Docker question, not a CapRover specific issue.

--------------------------------------------------------------------------

Please clearly state your question:

I am trying to deploy an application with the [haugene/transmission-openvpn](https://hub.docker.com/r/haugene/transmission-openvpn) image, which is based on the OpenVPN image. I tried to create an One-Click app template for it with the following code:

```json
{
  "captainVersion": "2",
  "displayName": "...",
  "description": "...",
  "dockerCompose": {
    "version": "3.3",
    "services": {
      "$$cap_appname": {
        "image": "haugene/transmission-openvpn",
        "volumes": [
          "$$cap_appname-data:/data",
          "/etc/localtime:/etc/localtime:ro"
        ],
        "restart": "always",
        "environment": {
          "CREATE_TUN_DEVICE": "true",
          "OPENVPN_PROVIDER": "...",
          "OPENVPN_CONFIG": "...",
          "OPENVPN_USERNAME": "...",
          "OPENVPN_PASSWORD": "..."
        },
        "logging": {
          "driver": "json-file",
          "options": {
            "max-size": "10m"
          }
        },
        "cap-add": "NET_ADMIN"
      }
    }
  },
  "instructions": {
    "start": "...",
    "end": "..."
  },
  "variables": []
}
```

However, I keep getting `ERROR: Cannot ioctl TUNSETIFF tun: Operation not permitted (errno=1)` in the logs after CapRover says the application is successfully deployed.

After searching with Google, I believe this is because of the container is not running with appropriate privilege, [so the `cap-add` argument is probably not successfully added](https://github.com/haugene/docker-transmission-openvpn/issues/488).

I can't find a detailed explanation in the CapRover documentation how exactly are the Docker Compose files mapped to one-click templates, so what I've done above is by following the one-click template examples. None of them uses `cap-add`.

I tried the following syntax variants, none working and all have this error.

```
"cap-add": "NET_ADMIN",
"cap-add": ["NET_ADMIN"]
```

**Comment by githubsaturn** [comment]
Please see https://github.com/moby/moby/issues/25885

TLDR: Docker swarm mode (what CapRover uses) doesn't support `cap-add`

On top of that, CapRover's one-click apps support only a subset of Docker Compose parameters. More details here: https://github.com/caprover/one-click-apps/blob/master/README.md

**Comment by davidmpa** [comment]
@githubsaturn Thank you for answer. Is there a way to use the Captain Definition file with `cap-add`, or it also doesn't support this (is it also based on swarm mode)?

(Sorry for the lame question)

**Comment by githubsaturn** [comment]
`captain-definition` won't be able to help either. CapRover services are entirely based on Docker swarm that allows for auto restart and better cluster management. Unfortunately, there are some limitations that come with Docker swarm and this is one of them. 

Alternatively, you can can manually deploy a Docker container on the same host. CapRover can peacefully live next to other containers that are created manually. It does not cause any interfere or problems.

**Comment by davidmpa** [comment]
Thank you.

**Comment by guilh22** [comment]
https://github.com/caprover/caprover/pull/2103
https://github.com/caprover/caprover-frontend/pull/151
